observatory-dags/observatory/dags/model.py

```diff
@@ -0,0 +1,1273 @@
+    journals = []
+    publishers = []
+    journal_id = 0
+    for i, _ in enumerate(range(n_publishers)):
+        n_journals_ = random.randint(1, 3)
+        journals_ = []
+        for _ in range(n_journals_):
+            journals_.append(Journal(journal_id, name=faker.company(), license=random.choice(LICENSES)))
+            journal_id += 1
+
+        publishers.append(Publisher(i, name=faker.company(), doi_prefix=doi_prefix, journals=journals_))
+        journals += journals_
+        doi_prefix += 1
+
+    # Fields of study
+    fields_of_study = []
+    fos_id_ = 0
+    for level in range(6):
+        for _ in range(n_fields_of_study_per_level):
+            n_words_ = random.randint(1, 3)
+            name_ = faker.sentence(nb_words=n_words_)
+            fos_ = FieldOfStudy(fos_id_, name=name_, level=level)
+            fields_of_study.append(fos_)
+            fos_id_ += 1
+
+    # Authors
+    authors = []
+    for i, _ in enumerate(range(n_authors)):
+        author = Author(i, name=faker.name(), institution=random.choice(institutions))
+        authors.append(author)
+
+    papers = []
+    for i, _ in enumerate(range(n_papers)):
+        # Random title
+        n_words_ = random.randint(2, 10)
+        title_ = faker.sentence(nb_words=n_words_)
+
+        # Random date
+        published_date_ = pendulum.strptime(faker.date(), "%Y-%m-%d").date()
+        published_date_ = pendulum.date(year=published_date_.year, month=published_date_.month, day=published_date_.day)
+
+        # Output type
+        output_type_ = random.choice(OUTPUT_TYPES)
+
+        # Pick a random list of authors
+        n_authors_ = random.randint(1, 10)
+        authors_ = random.sample(authors, n_authors_)
+
+        # Random funder
+        n_funders_ = random.randint(0, 3)
+        if n_funders_ > 0:
+            funders_ = random.sample(funders, n_funders_)
+        else:
+            funders_ = []
+
+        # Random publisher
+        publisher_ = random.choice(publishers)
+
+        # Journal
+        journal_ = random.choice(publisher_.journals)
+
+        # Random DOI
+        doi_ = make_doi(publisher_.doi_prefix)
+
+        # Random events
+        n_events_ = random.randint(0, 100)
+        events_ = []
+        for _ in range(n_events_):
+            range_months_ = random.randint(1, min(24, max((pendulum.date.today() - published_date_).in_months(), 1)))
+            date_end = published_date_.add(months=range_months_)
+            event_date_ = faker.date_between_dates(date_start=published_date_, date_end=date_end)
+            event_date_ = pendulum.datetime(year=event_date_.year, month=event_date_.month, day=event_date_.day)
+            events_.append(Event(source=random.choice(EVENT_TYPES), event_date=event_date_))
+
+        # Fields of study
+        n_fos_ = random.randint(1, 20)
+        level_0_index = 199
+        fields_of_study_ = [random.choice(fields_of_study[:level_0_index])] + random.sample(fields_of_study, n_fos_)
```

Comment:
this is possibly faster with .extend() on the list, or a chain (itertools).
https://python-forum.io/thread-5534.html
Will need to evaluate this use case to see if it's appropriate.

Reply:
Thanks Tuan, I've gone with .extend